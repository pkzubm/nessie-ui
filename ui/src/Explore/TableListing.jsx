/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, {useEffect, useState} from 'react';
import {Card, ListGroup, ListGroupItem} from "react-bootstrap";
import InsertDriveFileOutlinedIcon from '@material-ui/icons/InsertDriveFileOutlined';
import FolderIcon from '@material-ui/icons/Folder';
import ExploreLink from "./ExploreLink";
import PropTypes from 'prop-types';
import {api} from "../utils";


function groupItem(name, type, ref, path) {
  let icon = type === "CONTAINER" ? <FolderIcon/> : <InsertDriveFileOutlinedIcon/>;
  return (
    <ListGroupItem key={name}>
      <ExploreLink toRef={ref} path={path.concat(name)} type={type === "CONTAINER" ? "CONTAINER" : "OBJECT"}>
        {icon}{name}
      </ExploreLink>
    </ListGroupItem>
  )
}

function fetchKeys(ref, path, setKeys) {
  return api().getEntries({'ref':ref})
    .then(res => {
      return res.json();
    })
    .then((data) => {
      let keys = filterKeys(path, data.entries);
      setKeys(keys);
    }).catch(e => console.log(e));
}

// TODO: move this to server-side. Filter to keys relevant for this view.
function filterKeys(path, keys) {
  let containers = new Set();
  let filteredKeys = keys.map(key => key.name.elements)
    .filter(name => {
      if (name.length <= path.length) {
        return false;
      }

      // make sure all key values match the current path.
      return name.slice(0, path.length).every((v, i) => v.toLowerCase() === path[i].toLowerCase());
    })
    .map(name => {
      let ele = name[path.length];
      if (name.length > path.length + 1) {
        containers.add(ele);
      }
      return ele;
    });

  let distinctKeys = [...new Set(filteredKeys)];
  let distinctObjs = distinctKeys.map(key => {
    return {
      name: key,
      type: containers.has(key) ? "CONTAINER" : "TABLE"
    };
  })
  return distinctObjs;
}

function TableListing(props) {
  const [keys, setKeys] = useState([]);
  useEffect(() => {
    fetchKeys(props.currentRef, props.path, setKeys);
  }, [props.currentRef, props.path])

  return (
    <Card>
      <ListGroup variant={"flush"}>
        {keys.map(key => {return groupItem(key.name, key.type, props.currentRef, props.path);})}
      </ListGroup>
    </Card>
  );
}

TableListing.propTypes = {
  currentRef: PropTypes.string.isRequired,
  path: PropTypes.arrayOf(PropTypes.string).isRequired,
  currentLog: PropTypes.shape({
    committer: PropTypes.string.isRequired,
    hash: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    commitTime: PropTypes.number.isRequired
  })
}

TableListing.defaultProps = {
  currentRef: "main",
  path: [],
}

export default TableListing;
