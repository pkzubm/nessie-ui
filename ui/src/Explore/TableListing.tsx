/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { useEffect, useState } from "react";
import { Card, ListGroup, ListGroupItem } from "react-bootstrap";
import InsertDriveFileOutlinedIcon from "@material-ui/icons/InsertDriveFileOutlined";
import FolderIcon from "@material-ui/icons/Folder";
import ExploreLink from "./ExploreLink";
import {
  api,
  Branch,
  Content,
  ContentKey,
  Entry,
  GetMultipleContentsRequest,
  GetMultipleContentsResponse,
} from "../utils";
import { factory } from "../ConfigLog4j";
import { ContentView, EmptyMessageView } from "./Components";
import { useHistory, Redirect, useLocation } from "react-router-dom";
import { routeSlugs } from "./Constants";
import { Location, History } from "history";

const log = factory.getLogger("api.TableListing");

const groupItem = (
  key: Key,
  ref: string,
  path: string[]
): React.ReactElement => {
  const icon =
    key.type === "CONTAINER" ? <FolderIcon /> : <InsertDriveFileOutlinedIcon />;
  return (
    <ListGroupItem key={key.name}>
      <ExploreLink
        toRef={ref}
        path={path.concat(key.name)}
        type={key.type === "CONTAINER" ? "CONTAINER" : "OBJECT"}
      >
        {icon}
        {key.name}
      </ExploreLink>
    </ListGroupItem>
  );
};

const fetchKeys = (
  ref: string,
  path: string[],
  showContent: boolean
): Promise<void | Key[] | undefined> => {
  const newPath = showContent ? Array(path[0]) : path;
  return api()
    .getEntries({
      ref,
      namespaceDepth: newPath.length + 1,
      filter: `entry.namespace.matches('${newPath.join("\\\\.")}(\\\\.|$)')`,
    })
    .then((data) => {
      return data.entries?.map((e) => entryToKey(e));
    })
    .catch((e) => log.error("Entries", e as undefined));
};

const entryToKey = (entry: Entry): Key => {
  return {
    name: entry.name.elements[entry.name.elements.length - 1],
    type: entry.type === "UNKNOWN" ? "CONTAINER" : "TABLE",
    contentKey: entry.name,
  };
};

const fetchContent = (
  ref: string,
  hashOnRef: string | undefined,
  getMultipleContentsRequest: GetMultipleContentsRequest
): Promise<void | GetMultipleContentsResponse | undefined> => {
  return api()
    .getMultipleContents({
      ref,
      hashOnRef,
      getMultipleContentsRequest,
    })
    .then((data) => {
      return data;
    })
    .catch((e) => log.error("getMultipleContents", e as undefined));
};

interface Key {
  name: string;
  type: "CONTAINER" | "TABLE";
  contentKey: ContentKey;
}

interface ITableListing {
  currentRef: string;
  path: string[];
  branches: Branch[];
}

const TableListing = ({
  currentRef,
  path,
  branches,
}: ITableListing): React.ReactElement => {
  const [keys, setKeys] = useState<Key[]>([]);
  const [isRefNotFound, setRefNotFound] = useState(false);
  const [showContent, setShowContent] = useState(false);
  const [content, setcontent] = useState<Content>();
  const location = useLocation() as Location;
  const history = useHistory() as History;
  useEffect(() => {
    const modifiedPath = location.pathname && location.pathname.split("/")[1];
    const isContentPath = modifiedPath === routeSlugs.content;
    const keysFn = async () => {
      const fetched = await fetchKeys(currentRef, path, isContentPath);
      if (fetched) {
        setKeys(fetched);
      }
      setRefNotFound(fetched === undefined);
    };
    if (keys?.length === 0 || !isContentPath) void keysFn();
    setShowContent(isContentPath);
  }, [currentRef, path]);

  useEffect(() => {
    const getContent = async () => {
      const currentBranchDetail = branches.find((branch) => {
        return branch.name === currentRef;
      });
      if (keys?.length > 0) {
        const keyObj = keys[0].contentKey;
        const contentData = await fetchContent(
          currentRef,
          currentBranchDetail?.hash,
          { requestedKeys: [keyObj] }
        );
        const contentList =
          contentData?.contents && contentData.contents.length > 0
            ? contentData.contents
            : [];
        const contentDetails =
          contentList.length > 0 ? contentList[0].content : undefined;
        setcontent(contentDetails);
      }
    };
    if (showContent) {
      void getContent();
    }
  }, [showContent, keys]);

  useEffect(() => {
    if (showContent) {
      const listPath = location.pathname.split("/", 3).join("/");
      history.push(listPath);
    }
  }, [currentRef]);

  return isRefNotFound ? (
    <Redirect to="/notfound" />
  ) : (
    <Card>
      {!showContent ? (
        <ListGroup variant={"flush"}>
          {keys.map((key) => {
            return groupItem(key, currentRef, path);
          })}
        </ListGroup>
      ) : content ? (
        <ContentView tableContent={content} />
      ) : (
        <EmptyMessageView />
      )}
    </Card>
  );
};

TableListing.defaultProps = {
  currentRef: "main",
  path: [],
  branches: [],
};

export default TableListing;
