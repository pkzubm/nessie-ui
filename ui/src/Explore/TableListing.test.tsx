/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { BrowserRouter } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import TableListing from "./TableListing";
// tslint:disable-next-line:no-implicit-dependencies
import nock from "nock";

it("TableListing renders", async () => {
  const entry = {
    name: { elements: ["a"] },
    type: "UNKNOWN",
  };
  const scope = nock("http://localhost/api/v1")
    .get(
      "/trees/tree/main/entries?namespaceDepth=1&query_expression=entry.namespace.matches(%27(%5C%5C.%7C%24)%27)"
    )
    .reply(200, { token: "foo", entries: [entry] });

  const { getByText, asFragment } = render(
    <BrowserRouter>
      <TableListing currentRef={"main"} />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => getByText("a"));

  scope.done();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText("a").closest("a")).toHaveAttribute(
    "href",
    "/tree/main/a"
  );
});

it("TableListing renders object", async () => {
  const entry = {
    name: { elements: ["b"] },
    type: "TABLE",
  };
  const scope = nock("http://localhost/api/v1")
    .get(
      "/trees/tree/main/entries?namespaceDepth=1&query_expression=entry.namespace.matches(%27(%5C%5C.%7C%24)%27)"
    )
    .reply(200, { token: "foo", entries: [entry] });

  const { getByText, asFragment } = render(
    <React.StrictMode>
      <BrowserRouter>
        <TableListing currentRef={"main"} />
      </BrowserRouter>
    </React.StrictMode>
  );
  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => getByText("b"));

  scope.done();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText("b").closest("a")).toHaveAttribute(
    "href",
    "/contents/main/b"
  );
});
