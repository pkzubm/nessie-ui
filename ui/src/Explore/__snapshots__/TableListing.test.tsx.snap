// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableListing renders 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="list-group list-group-flush"
    />
  </div>
</DocumentFragment>
`;

exports[`TableListing renders 2`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item"
      >
        <a
          href="/tree/main/a"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
            />
          </svg>
          a
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TableListing renders object 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="list-group list-group-flush"
    />
  </div>
</DocumentFragment>
`;

exports[`TableListing renders object 2`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="list-group list-group-flush"
    >
      <div
        class="list-group-item"
      >
        <a
          href="/contents/main/b"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zM6 20V4h7v5h5v11H6z"
            />
          </svg>
          b
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
